scipy/sparse/csgraph/_flow.pyx

```diff
@@ -407,3 +431,207 @@ cdef ITYPE_t[:] _edmonds_karp(
+                    end = 0
+    return 0
+
+cdef ITYPE_t _augment_paths(
+        const ITYPE_t[:] edge_ptr, # IN
+        const ITYPE_t current, # IN
+        const ITYPE_t sink, # IN
+        const ITYPE_t[:] levels, # IN
+        const ITYPE_t[:] heads, # IN
+        const ITYPE_t[:] rev_edge_ptr, # IN
+        ITYPE_t[:] capacities, # IN/OUT
+        ITYPE_t[:] progress, # IN
+        ITYPE_t[:] flows, # OUT
+        ITYPE_t flow # OUT
+    ) nogil:
+    """Computes blocking flow in layered graph using depth first search.
+
+    Parameters
+    ----------
+    edge_ptr : memoryview of length :math:`|V| + 1`
+        For a given vertex ``v``, the edges whose tail is ``v`` are those between
+        ``edge_ptr[v]`` and ``edge_ptr[v + 1] - 1``.
+    current : int
+        The current node in the path.
+    sink : int
+        The sink vertex.
+    levels: memoryview of length :math:`|E|`
+        For a given vertex ``v``, ``levels[v]`` is the level of ``v`` in 
+        the layered graph of input graph.
+    heads : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``heads[e]`` is the head vertex of ``e``.
+    rev_edge_ptr : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``rev_edge_ptr[e]`` is the edge obtained by
+        reversing ``e``. In particular, ``rev_edge_ptr[rev_edge_ptr[e]] == e``.
+    capacities : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``capacities[e]`` is the capacity of ``e``.
+    progress: memoryview of length :math:`|E|`
+        For a given vertex ``v``, ``progress[v]`` is the index of the next
+        edge to be visited from ``v``.
+    flows : memoryview of length :math:`|E|`
+        The residual graph with respect to a maximum flow.
+    flow : int
+        Current maximum flow in the path.
+
+    Returns
+    -------
+    result_flow : int
+        The blocking flow in the layered graph.
+
+    """
+    if current == sink:
+        return flow
+    
+    cdef ITYPE_t dst_vertex, result_flow, e
+
+    if progress[current] == -1:
+        progress[current] = edge_ptr[current]
+    while progress[current] < edge_ptr[current + 1]:
+        e = progress[current]
+        dst_vertex = heads[e]
+        if (capacities[e] > 0 and 
+            levels[dst_vertex] == levels[current] + 1):
+            result_flow = _augment_paths(edge_ptr, dst_vertex, sink,
+                                         levels, heads, rev_edge_ptr,
+                                         capacities, progress, flows,
+                                         min(flow, capacities[e]))
+            if result_flow:
+                capacities[e] -= result_flow
+                capacities[rev_edge_ptr[e]] += result_flow
+                flows[e] += result_flow
+                flows[rev_edge_ptr[e]] -= result_flow
+                return result_flow
+        progress[current] += 1
+    return 0
+
+cdef ITYPE_t[:] _dinic(
+        ITYPE_t[:] edge_ptr,
+        ITYPE_t[:] tails,
```

Comment:
`tails` is unused (for Edmonds--Karp, it's used only during augmentation); in updating the preprocessing, I'll make it possible to not calculate `tails` as all.

Reply:
Addressed.